typing/typecore.ml

```diff
@@ -1285,7 +1406,8 @@ and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
       generalize_structure expected_ty;
       (* PR#7404: allow some_private_tag blindly, as it would not unify with
          the abstract row variable *)
-      if l = Parmatch.some_private_tag then assert (constrs <> None)
+      if l = Parmatch.some_private_tag
+      then assert (match mode with Normal -> false | Counter_example _ -> true)
```

Comment:
Is this a sign that `not_a_counter_example` should be called `typing_a_counter_example` so that you could use it here (as well as every where else, where you'd add `not` in the front of it)?

Reply:
I renamed `not_a_counter_example` into `construction_not_used_in_counterexamples`, because no one was able to guess yet that this is what I meant (but trying to be less verbose).

Reply:
Either you missed my point, or there's something that I do not understand.

Previously the code was sprinkled with `assert (constrs = None)` and `assert (constrs <> None)`, which was consistent.

With this PR, it uses `assert whathever_you_called_the_thing;` everywhere, except here where it uses this match, which seems to indicate that the concern is different.

My understanding was that the concern is not different, and that we're now just recomputing the information instead of using it. If that's the case, then I think it's a mistake.
But perhaps I just always misunderstood that part.

What do you say?

Reply:
> Previously the code was sprinkled with assert `(constrs = None)` and assert `(constrs <> None)`, which was consistent.

Consistent with what? You can't tell if it is consistent unless you look at the justification for why the assumption must hold. There may even be several asserts that check the same boolean condition, but with different justifications.

My best guess for what those `assert (constrs = None)` mean (all of them) is: this pattern-former is never used by the counter-example generator. If I wanted to formulate this justification as the negation of something, it would be the negation of "this pattern-former *can* be used by the counter-example generator", which is *not* equivalent to `constrs <> None` (because indeed even those pattern-formers need to be type-checked in Normal mode). 

I am not sure what the `assert (constrs <> None)` occurrence is about (it refers to an obscure PR and the PR number seems wrong), but my best guess is that it means that the condition `l = Parmatch.some_private_tag` can only be true if the pattern was generated in counter-example mode. I could name it `only_used_by_the_counter_example_generator`, and it's *not* the negation of the other condition. Because it's used only once, I just wrote the test inline: `assert (match mode with Normal -> false | Counter_example _ -> true)`.

Reply:
I see where you're coming from, and I know understand your choice of name for the boolean!

I read things slightly differently:
- `assert (constrs = None);`: "at this point, we cannot be refining a non exhaustiveness example"
- `assert (constrs <> None);`: "at this point, we must be refining a non exhaustiveness example"

And then, indeed, I would have to figure out, case by case, why that assertion holds.

---

Here you're proposing to understand the code slightly differently. Which might be a sensible thing to do.
Now that I understand the intent, I'd like to read over the code one more time, to check whether I agree with it, before giving a formal approval.
(I'll try to do that today, but it might have to wait until tomorrow morning)

Reply:
Thanks! I'm happy to see another external perspective of this and I'm willing to wait to get re-reviews. (I found other exciting OCaml things to spend my energy on in the meantime.)